utils/s3/client.cc

```diff
@@ -9,7 +9,11 @@
 #include <initializer_list>
 #include <memory>
 #include <stdexcept>
+#if __has_include(<rapidxml.h>)
```

Comment:
Yikes. Isn't it something that build configurator should take care of?

Reply:
see #15813. the name of header files are different. what we can do in `configure.py` or CMake side is to detect both of them, and conditionally define a macro. but i don't think the macro is necessary or very helpful here.

Reply:
OK. If go this route, let's then make it generic enough like

```c++
#if __has_include(A)
#include <A>
#else if __has_include(B)
#include <B>
#else
#error "No needed include"
#endif
```

Reply:
> OK. If go this route, let's then make it generic enough like
> 
> 
> 
> ```c++
> 
> #if __has_include(A)
> 
> #include <A>
> 
> #else if __has_include(B)
> 
> #include <B>

what is the merit of being "generic"? I'd avoid repeatings if possible?

> 
> #else
> 
> #error "No needed include"

compiler errors out twitch a better error message. and, if we run into this branch, it's a bug of the building system. why are we expect a bug here?

> 
> #endif
> 
> ```

Reply:
> what is the merit of being "generic"? I'd avoid repeatings if possible?

Well, _this_ place -- since we now know that rapidxml headers can be pretty "random" it seems natural to check them in locations we expect then print smth like "we tried, but failed :(" message

> and, if we run into this branch, it's a bug of the building system

Yes, that's why I proposed to locate the header location (or bail out if missing) in configure/cmake, not here

> why are we expect a bug here?

Because it turned out that there are distros that ship rapidxml via some other header other than what we expected. Now we extended our expectation, but no guarantee it now covers 100% of distros

Reply:
> > what is the merit of being "generic"? I'd avoid repeatings if possible?
> 
> 
> 
> Well, _this_ place -- since we now know that rapidxml headers can be pretty "random" it seems natural to check them in locations we expect then print smth like "we tried, but failed :(" message
> 
> 
> 
> > and, if we run into this branch, it's a bug of the building system
> 
> 
> 
> Yes, that's why I proposed to locate the header location (or bail out if missing) in configure/cmake, not here
> 
> 
> 
> > why are we expect a bug here?
> 
> 
> 
> Because it turned out that there are distros that ship rapidxml via some other header other than what we expected. Now we extended our expectation, but no guarantee it now covers 100% of distros

So, let's leave it to the building system, not the source file.

Reply:
I think it's okay to leave it here rather than to spread it to cmake where it's harder to understand for most people.

Reply:
@avikivity i don't understand your comment. "leave it", what does it mean here?

the building system ensures that we have either `rapidxml.h` or `rapidxml/rapidxml.hpp` can be found by the compiler, see #15813. 

i think the upper layers of the system have to always trust their underlying layers. in other words, we design by contract. so there is no need to add the extra handling here to be "safe" or "generic" or "to be easier to understand". on the contrary, it increases the complexity of the system, and blurs the boundary without good reasons or benefit. if we really wanted to go this way, the whole system would be bloated, as each layer have to repeat what has been done in its lower layers.

Reply:
I meant that #if __has_include concentrates the gunk in one place rather than spreading it around, and I prefer it this way. But I defer to you as the de-facto build system maintainer.